File: university-service/src/main/java/ua/com/foxminded/courseproject/service/TeacherServiceImpl.java
package ua.com.foxminded.courseproject.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import ua.com.foxminded.courseproject.dto.TeacherDto;
import ua.com.foxminded.courseproject.exceptions.TeacherConflictException;
import ua.com.foxminded.courseproject.exceptions.TeacherNotFoundException;
import ua.com.foxminded.courseproject.mapper.TeacherMapper;
import ua.com.foxminded.courseproject.repository.TeacherRepository;

import java.util.UUID;

@Service
public class TeacherServiceImpl implements PersonService<TeacherDto> {
    private TeacherMapper mapper;
    private TeacherRepository repository;

    @Autowired
    public TeacherServiceImpl(TeacherMapper mapper, TeacherRepository repository) {
        this.mapper = mapper;
        this.repository = repository;
    }

    @Override
    public TeacherDto findById(UUID id) {
        return mapper.toDto(repository.findById(id).orElseThrow(() -> new TeacherNotFoundException(id)));
    }

    @Override
    public Page<TeacherDto> findAll(Pageable pageable) {
        return repository.findAll(pageable).map(mapper::toDto);
    }

    @Override
    public TeacherDto save(TeacherDto teacher) {
        if (personExists(teacher)) {
            throw new TeacherConflictException(teacher);
        }
        return mapper.toDto(repository.save(mapper.toEntity(teacher)));
    }

    @Override
    public void delete(UUID id) {
        repository.delete(repository.findById(id).get());
    }

    @Override
    public Boolean personExists(TeacherDto personDto) {
        return repository.existsTeacherByFirstNameAndLastNameAndBirthDay(personDto.getFirstName(),
                personDto.getLastName(), personDto.getBirthDay());
    }
}

File: university-service/src/main/kotlin/ua/com/foxminded/courseproject/service/TeacherServiceImpl.kt
package ua.com.foxminded.courseproject.service

import java.util.UUID
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import ua.com.foxminded.courseproject.dto.TeacherDto
import ua.com.foxminded.courseproject.exceptions.TeacherConflictException
import ua.com.foxminded.courseproject.exceptions.TeacherNotFoundException
import ua.com.foxminded.courseproject.mapper.TeacherMapper
import ua.com.foxminded.courseproject.repository.TeacherRepository

@Service
class TeacherServiceImpl(
    private val mapper: TeacherMapper,
    private val repository: TeacherRepository
) : PersonService<TeacherDto> {
    override fun findById(id: UUID): Mono<TeacherDto> {
        return repository.findById(id)
            .switchIfEmpty(Mono.error(TeacherNotFoundException(id)))
            .map { mapper.toDto(it) }
    }

    override fun findAll(): Flux<TeacherDto> {
        return repository.findAll().map { mapper.toDto(it) }
    }

    override fun save(teacher: TeacherDto): Mono<TeacherDto> {
        return personExists(teacher).flatMap {
            if (it == true) {
                Mono.error(TeacherConflictException(teacher))
            } else {
                repository.save(mapper.toEntity(teacher)!!).map { mapper.toDto(it) }
            }
        }
    }

    override fun delete(id: UUID): Mono<Void> {
        return repository.findById(id).flatMap { repository.delete(it) }
    }

    override fun personExists(personDto: TeacherDto): Mono<Boolean> {
        return repository.existsTeacherByFirstNameAndLastNameAndBirthDay(
            personDto.firstName!!,
            personDto.lastName!!, personDto.birthDay!!
        )
    }
}
